aten/src/ATen/native/cpu/PowKernel.cpp

```diff
@@ -34,126 +36,104 @@ void pow_tensor_tensor_kernel(TensorIteratorBase& iter) {
   }
 }
 
-void pow_tensor_scalar_kernel(TensorIteratorBase& iter, const Scalar& exp_scalar) {
-  if (isFloatingType(iter.dtype())) {
-    const auto exp = exp_scalar.to<double>();
-    // Floating types allow AVX2 vector optimizations for pow/sqrt/rsqrt:
-    AT_DISPATCH_FLOATING_TYPES(iter.dtype(), "pow", [&]() {
+// Floating and complex types (minus Half) allow for optimized kernels
+template <typename scalar_t, typename cast_scalar_t, typename exp_scalar_t>
+void pow_tensor_scalar_optimized_kernel(TensorIteratorBase& iter, const exp_scalar_t exp) {
+  using Vec = Vec256<scalar_t>;
+  if (exp == 0.5) {
+    cpu_kernel_vec(iter,
+        [](scalar_t base) -> scalar_t {
+        	return std::sqrt(base);
+        },
+        [](Vec base) -> Vec { return base.sqrt(); }
+    );
+  } else if (exp == 2.0) {
+    cpu_kernel_vec(iter,
+        [](scalar_t base) -> scalar_t {
+        	return base * base;
+        },
+        [](Vec base) -> Vec { return base * base; }
+    );
+  } else if (exp == 3.0) {
+    cpu_kernel_vec(iter,
+        [](scalar_t base) -> scalar_t {
+        	return base * base * base;
+        },
+        [](Vec base) -> Vec { return base * base * base; }
+    );
+  } else if (exp == -0.5) {
+    cpu_kernel_vec(iter,
+        [](scalar_t base) __ubsan_ignore_float_divide_by_zero__ -> scalar_t {
+          return scalar_t(1.0) / std::sqrt(base);
+        },
+        [](Vec base) -> Vec { return base.rsqrt(); }
+    );
+  } else if (exp == -1.0) {
+    cpu_kernel_vec(iter,
+        [](scalar_t base) -> scalar_t {
+        	return scalar_t(1.0) / base;
+        },
+        [](Vec base) -> Vec { return base.reciprocal(); }
+    );
+  } else if (exp == -2.0) {
+    cpu_kernel_vec(iter,
+        [](scalar_t base) -> scalar_t {
+        	return scalar_t(1.0) / (base * base); },
+        [](Vec base) -> Vec { return (base * base).reciprocal(); }
+    );
+  } else {
+    cpu_kernel_vec(iter,
+        [=](scalar_t base) -> scalar_t {
+          return std::pow(base, static_cast<cast_scalar_t>(exp));
+        },
+        [=](Vec base) -> Vec {
+          return base.pow(static_cast<cast_scalar_t>(exp));
+        }
+    );
+  }
+}
+
+void pow_tensor_scalar_kernel(
+    TensorIteratorBase& iter,
+    const Scalar& exp_scalar) {
+  // prevent multiple calls to iter.common_dtype()
+  const auto dtype = iter.common_dtype();
+  if (dtype == ScalarType::Float || dtype == ScalarType::Double) {
+    AT_DISPATCH_FLOATING_TYPES(dtype, "pow", [&]() {
+      pow_tensor_scalar_optimized_kernel<scalar_t, double>(
+          iter, exp_scalar.to<double>());
+    });
+  } else if (isComplexType(dtype)) {
+    AT_DISPATCH_COMPLEX_TYPES(dtype, "pow", [&]() {
+      pow_tensor_scalar_optimized_kernel<scalar_t, scalar_t>(
```

Comment:
Same here `<scalar_t, c10::complex<double>, scalar_t>`

Reply:
Oops, this one is actually a bug I introduced

Reply:
> Same here `<scalar_t, c10::complex<double>, scalar_t>`

@heitorschueroff, that would be incorrect.
The third type is the exponent's type.
The second type should be scalar_t.
So it should be `<scalar_t, scalar_t, c10::complex<double>`

Reply:
@heitorschueroff, I think you didn't notice that I added `cast_scalar_t` as the second type, and not the third one in the template. In that case, can I rollback to using two types in function calls, as the Google style guide seems to be okay with it? Thanks!

Reply:
@imaginary-person I'm sorry for the confusion, when I experimented with this I added the cast type as the third value and I ended up reading your code with this bias. Yes, you can revert to leaving the last type implicit.